### ISSUE TITLE
[FEATURE] Action dependency injection

### ISSUE THREAD

**Issue by dbwiddis** [opensearch-project/opensearch-sdk-java#368]
### Is your feature request related to a problem?

Extension Actions are triggered by Rest Requests, or some actions trigger other actions.  Extensions are being designed to trigger other extension's actions as well.

These actions require many common components when they are initialized.  See for example the AnomalyDetector plugin actions (being migrated to an extension) ([search for `@Inject`](https://github.com/search?q=repo%3Aopensearch-project%2Fanomaly-detection%20%40Inject&type=code)).  These actions all have a constructor similar to this (number and ordering of objects differs).
```java
@Inject
public IndexAnomalyDetectorTransportAction(
    TransportService transportService,
    ActionFilters actionFilters,
    Client client,
    ClusterService clusterService,
    Settings settings,
    AnomalyDetectionIndices anomalyDetectionIndices,
    NamedXContentRegistry xContentRegistry,
    ADTaskManager adTaskManager,
    SearchFeatureDao searchFeatureDao
) { ... }
```

Instantiating these `Action` objects requires these objects to be passed to the constructor.  
### What solution would you like?

There are essentially three options for doing this:
1. Every time the object is instantiated, pass the appropriate objects.  
   - This leads to a lot of duplication
   - I'm currently using this approach in migration work, and after duplicating this approach 3 times, I decided there had to be a better way and was motivated to create this issue asking about the other 2 options.
2. Use an existing dependency injection framework such as [Google Guice](https://github.com/google/guice).  
   - This is what is currently implemented (and the reason for the `@Inject` annotation), with the dependency management in OpenSearch via the [`createComponents()`](https://github.com/opensearch-project/OpenSearch/blob/5273a86cc2630cca98681ff1f2e364ca7ea030a5/server/src/main/java/org/opensearch/plugins/Plugin.java#L120-L154) extension point.  
     - Note that the (pre-fork) comments mention "migration away from Guice", so I'm not sure whether that's a current intention or why it was (and if it still is) desired.
     - They are [read in during bootstrap](https://github.com/opensearch-project/OpenSearch/blob/5273a86cc2630cca98681ff1f2e364ca7ea030a5/server/src/main/java/org/opensearch/node/Node.java#L758) and then [injected into Guice](https://github.com/opensearch-project/OpenSearch/blob/5273a86cc2630cca98681ff1f2e364ca7ea030a5/server/src/main/java/org/opensearch/node/Node.java#L1096-L1102).
   - We would have to rewrite this to inject on the SDK side rather than the OpenSearch side
3. Create these objects in some map somewhere (probably on the `ExtensionsRunner`) with the class as the key and a singleton object as the value (possibly even lazily initialized).  
   - We'd only need access to the map (or ExtensionsRunner) passed to all objects, they could fetch what they need later.
   - This could accomplish the same essential purpose as the Guice option above, without an external dependency but easier to understand and more clear code.

### What alternatives have you considered?

See above.

### Do you have any additional context?

There are probably security-related concerns, but I believe most/all of those can be isolated to the `Client` we will be using, and we could exclude the Client from this implementation.

I'll admit I'm somewhat favoring option 3, but also realize the benefits of option 2 and wanted to get feedback before heading down that implementation road.

Interested in thoughts/preferences/feedback from anyone with insight, including @dblock @saratvemulapalli @owaiskazi19 @joshpalis @vibrantvarun @minalsha @peternied and anyone else these people think should chime in on here.

**Comment by dbwiddis** [comment]
Proposed "roll your own" concept (totally untested, subject to change but just jotting down ideas):
1. Define an `SDKAction` interface with an `execute()` method and `getInstance()` method required
2. Define an `SDKBaseAction` abstract class which implements `execute()` and calls the abstract `doExecute()`
3. On `ExtensionsRunner` create a `Map<Class<?>, Supplier<?>>` and populate it with:
  a. the arguments currently passed to `createComponents()` (class and getter methods on ExtensionsRunner)
  b. the objects returned from `createComponents()`
    i. TODO: preserve class identity, possibly just returning a map and merging the maps
    ii. The class constructors would fetch this map from the `ExtensionsRunner` 
    iii. On getting from the map, the suppliers would call the `getInstance()` methods which would create the singleton object.

**Comment by peternied** [comment]
@dbwiddis Sharing auto-generated client shims sounds pretty reasonable to me, I don't think it matters how they are made available,

However, it could be that you are referring to sharing java classes between extensions, this will exclude other language extensions from being able to interact with one another.  One of the huge values of the Extensions SDK this is decoupling the implementation language of OpenSearch from the language of the extension.  What do you think of this trade-off?

**Comment by peternied** [comment]
From a security standpoint, if we are talking about sharing java classes, my extension could suddenly run untrusted code 😱 This will make handshaking / trust relationships to shore up these risks is very complicated.

**Comment by dbwiddis** [comment]
> However, it could be that you are referring to sharing java classes between extensions

Nope, in this case all these objects are self-contained in the extension, but need to be shared around multiple objects in the extension.  That has corollaries in other languages.

As for the SDK-provided objects (XContent registry, Settings, SDKClusterService, etc.) these will be API endpoints that will need to be implemented in another language but not the same as the open-ended "Actions" implementations.

**Comment by peternied** [comment]
Thanks for the clarification @dbwiddis

**Comment by joshpalis** [comment]
Thanks @dbwiddis for writing this up, I vote using Google Guice. As you have noted with how you're currently handling migration work, we can see that there is a certain amount of overhead when trying to rewire a plugin's transport action implementation with the SDK. I think we'll see the benefit of using Google Guice down the line once we start migrating other plugins as extensions, since plugin transport actions are already using Google Guice.

**Comment by vibrantvarun** [comment]
We can think of implementing dagger [https://dagger.dev/](https://dagger.dev/). It is lightweight and compile time only.

**Comment by dbwiddis** [comment]
Great suggestion @vibrantvarun!  Looks like a lightweight version of Guice with a subset of capability (but since we're only instantiating singletons that's all we need I think).  Since it's compile time it doesn't have to use (slow) reflection so should definitely offer better performance.

**Comment by owaiskazi19** [comment]
@dbwiddis what option did we move with? Based on our use case I'm fine with either Guice or Dagger.

**Comment by dbwiddis** [comment]
Haven't started either. Currently tracing through the code of how it's done with Guice and seeing if there are corresponding (and more efficient) ways to do it in Dagger

**Comment by dbwiddis** [comment]
The method we want to put on `SDKRestClient` is from `org.opensearch.client.OpenSearchClient` interface (not to be confused with the same named client on the Java client:
```java
<Request extends ActionRequest, Response extends ActionResponse> void execute(
    ActionType<Response> action,
    Request request,
    ActionListener<Response> listener
);
```
This is implemented on the `AbstractClient`:
```java
@Override
public final <Request extends ActionRequest, Response extends ActionResponse> void execute(
    ActionType<Response> action,
    Request request,
    ActionListener<Response> listener
) {
    doExecute(action, request, listener);
}
```
where `doExecute()` further is implemented by `NodeClient`:
```java
@Override
public <Request extends ActionRequest, Response extends ActionResponse> void doExecute(
    ActionType<Response> action,
    Request request,
    ActionListener<Response> listener
) {
    // Discard the task because the Client interface doesn't use it.
    executeLocally(action, request, listener);
}
```
leading to
```java
    public <Request extends ActionRequest, Response extends ActionResponse> Task executeLocally(
        ActionType<Response> action,
        Request request,
        ActionListener<Response> listener
    ) {
        return transportAction(action).execute(request, listener);
    }
```

All this is relatively straightforward, but here's where the dependency injection starts.

The `transportAction(action)` looks up the action (via `actions.get(action)`) from a map which is injected via Guice, taking a singleton instance of `ActionType<Response extends ActionResponse>`.  These are all the stub files you see that only have a name and singleton instance, such as `DeleteModelAction`.  If we want to keep the same `execute()` implementation for migration, we need to keep these singleton classes.

The map is initialized in `Node.java` line 1106:
```java
client.initialize(injector.getInstance(new Key<Map<ActionType, TransportAction>>() { ... })
```
Entries are added to the map via `getActions()` which returns a new `ActionHandler` with parameters for the same singleton class as before, and the transport action class to be registered:
```java
new ActionHandler<>(ReindexAction.INSTANCE, TransportReindexAction.class)
```
Again, if we want to keep the same implementation for migration, we need to keep these same pairs of classes.

On initialization the `ActionModule` calls `actions = setupActions(actionPlugins);` which creates an action registry and individually registers actions from the Node (`actions.register(MainAction.INSTANCE, TransportMainAction.class);`) but more interestingly registers them from the plugins:
```java
actionPlugins.stream().flatMap(p -> p.getActions().stream()).forEach(actions::register);
```
where `register` maps the name (from the instance) to the handler (with the instance and transport class)
```java
public void register(ActionHandler<?, ?> handler) {
    register(handler.getAction().name(), handler);
}
```
There's some magic behind the scenes where Guice in `org.opensearch.common.inject.spi.Elements` initializes modules, configuring them via `configure()` where this code is called:
```java
MapBinder<ActionType, TransportAction> transportActionsBinder = MapBinder.newMapBinder(
    binder(),
    ActionType.class,
    TransportAction.class
);
for (ActionHandler<?, ?> action : actions.values()) {
    // bind the action as eager singleton, so the map binder one will reuse it
    bind(action.getTransportAction()).asEagerSingleton();
    transportActionsBinder.addBinding(action.getAction()).to(action.getTransportAction()).asEagerSingleton();
    for (Class<?> supportAction : action.getSupportTransportActions()) {
        bind(supportAction).asEagerSingleton();
    }
}
```
All this is a complex way of binding the action instance to the transport class.

So the TLDR here is that as long as we can do the binding bit, we can replicate most of this.

**Comment by dbwiddis** [comment]
Have spent a few hours playing with Dagger.  Got it working but it has somewhat of a fatal flaw.  It absolutely requires the `@Component` annotation of an interface, from which it generates the necessary bits for injection.   Of course it worked on the command line:
```
$ gradle dagger

> Task :dagger
It works!
```
But the blocker for me is that my IDE (Eclipse) doesn't play nice with the Dagger-generated classes, specifically my factory class prepended with "Dagger".
<img width="356" alt="Screenshot 2023-02-07 at 10 43 25 PM" src="https://user-images.githubusercontent.com/9291703/217454356-b2b571fe-ba46-4938-9294-2abd05170c39.png">

Apparently there are configurations to make it work with Maven, but not Gradle.

Fortunately with the guided tour above I'm pretty sure I can figure out how to get it working with Guice.

**Comment by dbwiddis** [comment]
> There's some magic behind the scenes where Guice in `org.opensearch.common.inject.spi.Elements` initializes modules, configuring them via `configure()`

This looks like the entry point we need in extensions:
 - A class implementing `Module` needs to be created
 - The `configure()` method will implement the same code binding `getActions()` as objects
 - That method will be called by ExtensionsRunner during startup
 - We'll also need to bind all the other SDK-provided components (client, etc.)

**Comment by dbwiddis** [comment]
> > There's some magic behind the scenes where Guice in `org.opensearch.common.inject.spi.Elements` initializes modules, configuring them via `configure()`
> 
> This looks like the entry point we need in extensions:

This has been implemented in #436 as part of the `ExtensionsRunner` constructor.  No `Module` class is created directly but it's automatically used as a parameter to the `Guice.inject()` call in a lambda. In that code the SDK-provided components are bound (the lambda is executing the `configure(Binder b)` functional interface) and in addition any objects returned via `createComponents()` are also bound.

The next step is to bind the actions in a `MapBinder`.  The approach will parallel what currently happens in the `ActionModule` in OpenSearch:
 - read the `getActions()` extension point using same syntax as `ActionPlugin` implementation
 - inject the actions with the action class as the key and action handler as the instance
 - register the actions locally in a map with a String key based on the action name
 - register the actions remotely in OpenSearch via `org.opensearch.extensions.RegisterTransportActionsRequest`

I'm also looking to implement `Client.execute()` which internally takes a `ActionType<Response> action` and looks up the (injected) handler to execute its `doExecute()` method.  There's a `transportAction(action)` method that does this lookup that I'll make sure is available somewhere.

CC @joshpalis @vibrantvarun @saratvemulapalli for JS integration
